go/libkb/shared_dh.go

```diff
@@ -0,0 +1,225 @@
+package libkb
+
+import (
+	"fmt"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+	context "golang.org/x/net/context"
+	"sync"
+)
+
+// SharedDHKeyGeneration describes which generation of DH key we're talking about.
+// The sequence starts at 1, and should increment every time the shared DH key
+// rotates, which is everytime a device is revoked.
+type SharedDHKeyGeneration int
+
+type SharedDHSecretKeyBox struct {
+	Generation  SharedDHKeyGeneration `json:"generation"`
+	Box         string                `json:"box"`
+	ReceiverKID keybase1.KID          `json:"receiver_kid"`
+}
+
+type sharedDHSecretKeyBoxes struct {
+	Boxes []SharedDHSecretKeyBox `json:"boxes"`
+}
+
+// SharedDHKeyMap is a map of Generation numbers to
+// DH private keys, for decrypting data.
+type SharedDHKeyMap map[SharedDHKeyGeneration]NaclDHKeyPair
+
+// SharedDHKeyring holds on to all versions of the Shared DH key.
+// Generation=0 should be nil, but all others should be present.
+type SharedDHKeyring struct {
+	Contextified
+	sync.Mutex
+	uid         keybase1.UID
+	generations SharedDHKeyMap
+}
+
+// NewSharedDHKeyring makes a new SharedDH keyring for a given UID.
+func NewSharedDHKeyring(g *GlobalContext, uid keybase1.UID) *SharedDHKeyring {
+	return &SharedDHKeyring{
+		Contextified: NewContextified(g),
+		uid:          uid,
+		generations:  make(SharedDHKeyMap),
+	}
+}
+
+// CurrentGeneration returns what generation we're on. The version possible
+// Version is 1. Version 0 implies no keys are available.
+func (s *SharedDHKeyring) CurrentGeneration() SharedDHKeyGeneration {
+	return SharedDHKeyGeneration(len(s.generations))
```

Comment:
Needs a lock to be public. But also needs to not deadlock inside other `*Locked` functions, so could add `currentGenerationLocked`.